src/libsystemd/sd-hwdb/sd-hwdb.c

```diff
@@ -164,10 +164,16 @@ static int hwdb_add_property(sd_hwdb *hwdb, const struct trie_value_entry_f *ent
                 entry2 = (const struct trie_value_entry2_f *)entry;
                 old = ordered_hashmap_get(hwdb->properties, key);
                 if (old) {
-                        /* on duplicates, we order by filename and line-number */
-                        r = strcmp(trie_string(hwdb, entry2->filename_off), trie_string(hwdb, old->filename_off));
-                        if (r < 0 ||
-                            (r == 0 && entry2->line_number < old->line_number))
+                        /* On duplicates, we order by filename and line-number. We rely
+                         * on the fact that the filenames in the hwdb are stored in the
+                         * order of priority (higher later), because they are *processed*
+                         * in the order of priority. So we compare the indices to determine
+                         * which file had higher priority. Comparing the strings alphabetically
+                         * would be useless, because those are full paths, and e.g.
+                         * /usr/lib would sort after /etc, even though it has lower priority.
+                         */
+                        if (entry2->filename_off < old->filename_off ||
+                            (entry2->filename_off == old->filename_off && entry2->line_number < old->line_number))
```

Comment:
That does not work. The strbuf implementation merges shared prefixes, so even if the strings are inserted in-order, there is no guarantee they end up with ascending offsets.